src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/typeresolution.cpp

```diff
@@ -0,0 +1,653 @@
+#pragma once
+
+#include <RoMetadataApi.h>
+#include <hstring.h>
+#include <atlbase.h>
+#include <windows.foundation.h>
+#include <Windows.h>
+#include <wil/result.h>
+#include <wil/resource.h>
+#include <wil/filesystem.h>
+#include <Synchapi.h>
+#include "TypeResolution.h"
+#include "catalog.h"
+
+namespace UndockedRegFreeWinRT
+{
+    BOOL CALLBACK GetProcessExeDirInitOnceCallback(
+        _Inout_     PINIT_ONCE,
+        _Inout_opt_ PVOID,
+        _Out_opt_   PVOID*)
+    {
+        wil::unique_process_heap_string localExePath;
+        HRESULT hr = wil::GetModuleFileNameW(nullptr, localExePath);
+        if (FAILED_LOG(hr))
+        {
+            SetLastError(hr);
+            return FALSE;
+        }
+
+        // Modify the retrieved string to truncate the actual exe name and leave the containing directory path. This API
+        // expects a buffer size including the terminating null, so add 1 to the string length.
+        hr = PathCchRemoveFileSpec(localExePath.get(), wcslen(localExePath.get()) + 1);
+        if (FAILED_LOG(hr))
+        {
+            SetLastError(hr);
+            return FALSE;
+        }
+
+        g_cachedProcessExeDir = std::move(localExePath);
+        return TRUE;
+    }
+
+    // Returned string is cached globally, and should not be freed by the caller.
+    HRESULT GetProcessExeDir(PCWSTR* path)
+    {
+        *path = nullptr;
+        static INIT_ONCE ProcessExeDirInitOnce = INIT_ONCE_STATIC_INIT;
+
+        RETURN_IF_WIN32_BOOL_FALSE(InitOnceExecuteOnce(&ProcessExeDirInitOnce, GetProcessExeDirInitOnceCallback, nullptr, nullptr));
+
+        // The cache has been successfully populated by the InitOnce, so we can just use it directly.
+        *path = g_cachedProcessExeDir.get();
+        return S_OK;
+    }
+
+    HRESULT FindTypeInMetaDataFile(
+        __in IMetaDataDispenserEx* pMetaDataDispenser,
```

Comment:
If you're going to use SAL annotations, use the publicly documented updated annotations.
```
__in -> _In_
__deref_opt_out_opt -> _COM_Outptr_opt_result_maybenull_
__out_opt -> _Out_opt__
```

Reply:
Are SAL annotations still useful here in github though?